Showing connection modal when oauth is needed

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-549
Figma nFI8jCCAQnd19PNwScqVU6-fi

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

As requested by our design team, this PR changes how the Account Status section appears for merchants who did not connect via OAuth like below:
<img width="766" height="381" alt="Screenshot 2025-09-19 at 16 48 01" src="https://github.com/user-attachments/assets/6eb0cca6-9ca6-4053-81a2-97ebdbaa9beb" />

It also adds a new notice on top of the settings page:
<img width="577" height="109" alt="Screenshot 2025-09-19 at 18 06 03" src="https://github.com/user-attachments/assets/c383fe61-17fc-4d55-b231-dd17160b94df" />

The goal is to incentivise more merchants to connect via OAuth to improve security.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout to this branch on your test environment (`update/showing-connection-modal-when-oath-is-needed`)
- Set up your store to behave as if it did not connect via OAuth by removing the connection_type (or test_connection_type attr from the settings:
```
wp option patch delete woocommerce_stripe_settings test_connection_type
```
- Confirm you can see the different versions of the connect status section and the new notice on top, and that both look like the preview above.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Showing connection modal when oauth is needed

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

wjrosa committed:
Adding the info icon

wjrosa committed:
Improving tests PHP docs

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

diegocurbelo reviewed (commented):
The code looks good and it works as described in the PR.... but I think the UX might be confusing.

When the merchant is using manual/shared keys the gateway can still be used for purchases but settings are not available... maybe we can load the onboarding screen by default, but also keep the "Payment Methods" and "Settings" tabs? 
Or change the `Connect an account" buttons to the same "Re-authenticate" button we use in the current CTA... and maybe the tittle, currently is "Get started with Stripe" (both both onboarding and re-connections)

**BEFORE**
<img width="1720" height="997" alt="Screenshot 2025-09-15 at 6 30 33 PM" src="https://github.com/user-attachments/assets/2011b018-5965-4c11-b1c6-81f5218a33ad" />

**AFTER**
<img width="1719" height="723" alt="Screenshot 2025-09-15 at 6 29 15 PM" src="https://github.com/user-attachments/assets/f9899681-1f25-42cb-bc34-470996f57da8" />


--

Also, there is a small error in the testing instructions:

>Set up your store to behave as if it did not connect via OAuth. You can do it by either:
-- Removing your keys directly in the database after connecting

This actually disconnects the plugin (the gateway can't be used for purchases), to simulate a non-connected merchant, we need to remove the `connection_type` (or `test_connection_type` attr from the settings:
```
wp option patch delete woocommerce_stripe_settings test_connection_type
```

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

daledupreez commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -41,11 +41,12 @@ public function __construct() {
 	 * @version 4.0.0
 	 */
-	public function add_admin_notice( $slug, $class, $message, $dismissible = false ) {
+	public function add_admin_notice( $slug, $class, $message, $dismissible = false, $icon_img_url = '' ) {
```
Is it worth making this argument type explicit?
```suggestion
	public function add_admin_notice( $slug, $class, $message, $dismissible = false, string $icon_img_url = '' ) {
```

wjrosa replied:
I have removed the icon following the latest suggestion from Orcun above. So, this parameter is no longer needed.

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

wjrosa commented:
Thanks for the review, Diego!

> .... but I think the UX might be confusing.

Yeah, I agree! @orcungogus could you please advise here?

Another suggestion would be to update the existing "Account details" block instead. Changing the "Configure connection" button to "Re-connect" and some other parts of it:

<img width="1052" height="325" alt="Screenshot 2025-09-17 at 10 40 43" src="https://github.com/user-attachments/assets/dd0ed345-b54e-462e-a6c5-1038f8a39f69" />

> Also, there is a small error in the testing instructions:

Oops. Thanks for pointing that out

orcungogus commented:
> When the merchant is using manual/shared keys the gateway can still be used for purchases but settings are not available...

So, these merchants are still authenticated (with the old method) and they can still make use of the _Payment methods_ and _Settings_ tabs, correct? If so I like Wesley’s suggestion of taking over the account details CTA, just use "Reauthenticate" for that as well. In addition to an inline notice shown there, we should also show an alert notice on both _Payment methods_ and _Settings_ tabs, what do you think?

<img width="1440" height="900" alt="Stripe - Reconnect Notice - Payment Methods" src="https://github.com/user-attachments/assets/a6dda290-c0c7-41f3-a76f-37a309f5ecac" />
<img width="1440" height="1200" alt="Stripe - Reconnect Notice - Settings" src="https://github.com/user-attachments/assets/6ff0b6e8-ef7e-4518-8447-bca6ebc738b6" />

You can find these comps [here in Figma](https://www.figma.com/design/nFI8jCCAQnd19PNwScqVU6/Stripe-Payments--External-?node-id=2348-10772&t=kLcLsumeKpMHuQW6-4).

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

wjrosa committed:
Reverting the logic to display the connection modal

wjrosa committed:
Reconnect notice inside the connection status section

wjrosa committed:
Implementing the reconnect notice

wjrosa committed:
Unit test

wjrosa committed:
Changing the style of the new top notice

wjrosa committed:
Enabling styles when OC is not enabled

wjrosa committed:
Enabling styles when OC is not enabled

wjrosa commented:
Thanks, Orcun! I think it looks better 👍 . I have updated the implementation in my latest commits to follow the new designs. And also updated the PR description with the previews. I did not include the "dismiss" link, as we already have the button on the top right.

wjrosa committed:
Fix tests

wjrosa committed:
Unit test

wjrosa committed:
Renaming test file

daledupreez commented on client/settings/account-details/__tests__/index.test.js:
```diff
@@ -89,4 +93,29 @@ describe( 'AccountDetails', () => {
+		render( <AccountDetails /> );
+
+		expect(
+			screen.queryByText( /reconnect your Stripe account/i )
```
Is it worth using the longer text to be clearer as to what we're looking for? That also means we should also get clearer signals when there are failures.

wjrosa replied:
Sure. I just did it that way due to how the previous tests were written + Copilot suggestion. Updated it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/b16227d1fb37415c83e2e32f53ed78cc3c295bdf

daledupreez commented on client/settings/account-details/reconnect-notice.js:
```diff
@@ -0,0 +1,35 @@
+	return (
+		<NoticeContainer>
+			<span data-testid="help">
+				<Icon icon={ info } size="18" />
```
I believe the design calls for the `caution` icon rather than `info`.

wjrosa replied:
I don't see the `caution` icon as part of the package 🤔  (as `alert` as it was named in Figma):
<img width="670" height="59" alt="Screenshot 2025-09-23 at 14 26 23" src="https://github.com/user-attachments/assets/6e15695c-3228-4bf5-98f9-ae5f15b4a0c7" />

`info` was the closest I could find. I can export the icon as SVG and use it instead if that's better

daledupreez replied:
I think this is a versioning/dependency issue -- the `caution` icon was added in late 2024 as part of https://github.com/WordPress/gutenberg/pull/66555.

wjrosa replied:
Nice find! I have upgraded the package and replaced the icon in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/0b9e97ba0625d0ecf33eec01e4fc504461525993

wjrosa replied:
Preview:
<img width="727" height="74" alt="Screenshot 2025-09-23 at 17 47 08" src="https://github.com/user-attachments/assets/7b879fc1-d246-4f22-b885-0134c66bd21b" />

daledupreez commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -73,7 +74,7 @@ public function admin_notices() {
 
 		foreach ( (array) $this->notices as $notice_key => $notice ) {
-			echo '<div class="' . esc_attr( $notice['class'] ) . '" style="position:relative;">';
+			echo '<div class="' . esc_attr( $notice['class'] ) . '" style="position:relative; overflow: auto;">';
```
Given that this will impact all notices, I think we need some explicit documentation for why its needed, and what we should look for in other notices.

wjrosa replied:
I decided to keep the `overflow` style only when there are actions to render. And also added a comment about it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/b66a54db8b563cfc76cc838ab4a67e8fbb922c58

daledupreez commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -393,6 +411,29 @@ public function stripe_check_environment() {
+				$oauth_required = $needs_live_oauth || $needs_test_oauth;
+				if ( $oauth_required ) {
+					$message = __( 'Please reconnect to continue using Stripe and avoid disruptions on your store.', 'woocommerce-gateway-stripe' );
+					$link    = esc_url( admin_url( 'admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings&highlight=account-details' ) );
```
As a minor note, the `highlight` option doesn't seem to support the account details section. Not sure if that's something you want to tackle in this PR, or simply remove the query parameter for now.

wjrosa replied:
Oops, I forgot to implement that 🤦  . Added it now in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/39483773089edae6e49da5fbf42f91d988ec0152

daledupreez commented on assets/images/info.svg:
It looks like we're using the `info` icon from `@wordpress/icons`, so I don't believe we need this file.

wjrosa replied:
I can use it in the above if you think that's better. I will wait for this decision before removing it

wjrosa replied:
Deleted it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/48270ef11e79416dda97072de0d9accb2bcfb7ad

daledupreez commented on tests/phpunit/Admin/WC_Stripe_Admin_Notices_Test.php:
```diff
@@ -76,13 +127,30 @@ public function test_no_notices_are_shown_when_stripe_is_not_enabled() {
 	}
 
 	/**
+	 * Test that the correct notices are shown in all scenarios.
```
Thanks for all the PHPDocs!

daledupreez commented on tests/phpunit/Admin/WC_Stripe_Admin_Notices_Test.php:
```diff
@@ -109,107 +189,27 @@ public function test_correct_stripe_notices_are_shown_in_all_scenarios( $options
+	 *
+	 * @return array
+	 */
+	public function options_to_notices_map(): array {
```
Good call on co-locating the data provider and test runner functions.

daledupreez reviewed (commented):
Everything is working as expected, but we are not using the correct icon, and we're adding an unused SVG file. I am also concerned about the additional style rule being applied to all notices.

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

wjrosa committed:
Improve unit test

wjrosa committed:
Making the overflow style change strict to notices with actions + comment

wjrosa committed:
Correctly handling the highlight of account details in the frontend

diegocurbelo reviewed (approved):
Thanks for the changes @wjrosa. It works as expected!

wjrosa committed:
Upgrading wordpress/icons package + replacing the notice icon with the caution icon

wjrosa committed:
Removing the unused icon

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

daledupreez commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -73,7 +74,14 @@ public function admin_notices() {
 
 		foreach ( (array) $this->notices as $notice_key => $notice ) {
-			echo '<div class="' . esc_attr( $notice['class'] ) . '" style="position:relative;">';
+			$has_actions = count( $notice['actions'] ) > 0;
```
Just to be safe and prevent warnings, I think we should make sure that `$notice['actions']` is defined and countable.
```suggestion
			$has_actions = count( $notice['actions'] ?? [] ) > 0;
			// Or even something like this:
			$notice_actions = $notice['actions'] ?? [];
			$has_actions   = is_array( $notice_actions ) && [] !== $notice_actions;
```

wjrosa replied:
Alright. Refactored it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/d56e1ea6699e0f2f5b2f3b82a6c5e91eab30776b

daledupreez commented on includes/admin/class-wc-stripe-admin-notices.php:
```diff
@@ -73,7 +74,14 @@ public function admin_notices() {
 		foreach ( (array) $this->notices as $notice_key => $notice ) {
-			echo '<div class="' . esc_attr( $notice['class'] ) . '" style="position:relative;">';
+			$has_actions = count( $notice['actions'] ) > 0;
+			$div_style   = 'position:relative;';
```
Super-tiny nit: I think it would be easier to read if we swapped these two lines so the unconditional set of `$div_style` occurs before we start running any logic related to `$has_actions`.
```suggestion
			$div_style   = 'position:relative;';
			$has_actions = count( $notice['actions'] ) > 0;
```

wjrosa replied:
Swapped them in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4669/commits/9ea55bc0a27e8650fb6ae20d13726b16b6064f5e

daledupreez commented on package.json:
```diff
@@ -44,7 +44,7 @@
     "@wordpress/hooks": "3.2.0",
     "@wordpress/i18n": "^4.2.1",
-    "@wordpress/icons": "^5.0.1",
+    "@wordpress/icons": "10.31.0",
```
Do we want this to be a `^`-prefixed dependency? Or should we keep it as an explicit dependency?
```suggestion
    "@wordpress/icons": "^10.31.0",
```

wjrosa replied:
Let's set the prefixed version 👍

daledupreez commented on client/settings/payment-settings/account-details-section.js:
```diff
@@ -93,15 +94,32 @@ const AccountSettingsDropdownMenu = ( {
 			<CardHeader>
 				<HeaderDetails>
-					<h4>
+					<h4 ref={ headingRef }>
```
My only minor nit is that when I click on the notice CTA from the main WordPress dashboard, the auto-scroll moves me to a weird scroll location. I am not sure that we can improve this much.

<img width="1088" height="308" alt="Screenshot 2025-09-24 at 15 28 06" src="https://github.com/user-attachments/assets/8d42ffd1-b25b-43ea-aac1-354054def0e2" />

wjrosa replied:
Yeah. If I set the ref to the button, the settings bar covers it:
<img width="979" height="211" alt="Screenshot 2025-09-24 at 11 26 17" src="https://github.com/user-attachments/assets/a6b42941-9e01-47ba-a649-6733a96469c4" />

wjrosa committed:
Merge branch 'develop' into update/showing-connection-modal-when-oath-is-needed

daledupreez reviewed (approved):
Changes are looking good to me and testing well. My only issue was with the auto-scroll when coming from the main wp-admin dashboard when using Chrome.

wjrosa committed:
Update package.json

Co-authored-by: daledupreez <dale.du.preez@automattic.com>

wjrosa committed:
Package lock update

wjrosa committed:
Update has actions verification logic

wjrosa committed:
Swapping has actions and div style variables
